Add support for Atlasssian Bamboo Build Server

krippz opened the pull request:
Adding Atlasssian Bamboo Build Server via BambooProvider,  Closes #544

devlead commented:
@krippz this PR has a few merge conflicts and therefore won't build on our CI, could you please rebase against latest version of Cake.

krippz commented:
So I've fixed some stuff... :)

devlead commented on src/Cake.Core/CakeEnvironment.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
```
This file only has an encoding change, so shouldn't be part of the PR.

krippz replied:
sorry about that, i'll have done some reorganizing. will try and update PR later tonight

devlead commented:
Hi @krippz,

I've now reviewed and tested this on Bamboo and in short everything seems to work great. As you stated there are some code duplication but I rather see we address that in an later separate refactor PR.

Tested both standard build
And it picks up build, scm and project info as it should.
Log: https://cakebuild.atlassian.net/builds/download/CAKE-KRIP-JOB1/build_logs/CAKE-KRIP-JOB1-1.log

Object graph:
```json
{
  "IsRunningOnBamboo": true,
  "Environment": {
    "ShortJobKey": "JOB1",
    "ShortJobName": "Default Job",
    "ShortPlanKey": "KRIP",
    "ShortPlanName": "krippz-bamboo-build",
    "Project": {
      "PlanKey": "CAKE-KRIP"
    },
    "Build": {
      "Folder": "C:\\build\\CAKE-KRIP-JOB1",
      "Number": 1,
      "BuildKey": "CAKE-KRIP-JOB1",
      "ResultKey": "CAKE-KRIP-JOB1-1",
      "ResultsUrl": "https://cakebuild.atlassian.net/builds/browse/CAKE-KRIP-JOB1-1",
      "CustomBuild": {
        "IsCustomBuild": false,
        "RevisonName": ""
      }
    },
    "Repository": {
      "Scm": "git",
      "Name": "Devlead/Cake",
      "Branch": "krippz-bamboo-build",
      "Commit": {
        "RepositoryRevision": "a7c84e987b721b1832e5b7f8671e32d530639211"
      }
    }
  }
}
```

Also tested custom build with a specific reviosion and it picked up that as webb
Log: https://cakebuild.atlassian.net/builds/download/CAKE-KRIP-JOB1/build_logs/CAKE-KRIP-JOB1-2.log

Object graph;
```json
{
  "IsRunningOnBamboo": true,
  "Environment": {
    "ShortJobKey": "JOB1",
    "ShortJobName": "Default Job",
    "ShortPlanKey": "KRIP",
    "ShortPlanName": "krippz-bamboo-build",
    "Project": {
      "PlanKey": "CAKE-KRIP"
    },
    "Build": {
      "Folder": "C:\\build\\CAKE-KRIP-JOB1",
      "Number": 2,
      "BuildKey": "CAKE-KRIP-JOB1",
      "ResultKey": "CAKE-KRIP-JOB1-2",
      "ResultsUrl": "https://cakebuild.atlassian.net/builds/browse/CAKE-KRIP-JOB1-2",
      "CustomBuild": {
        "IsCustomBuild": true,
        "RevisonName": "a7c84e987b721b1832e5b7f8671e32d530639211"
      }
    },
    "Repository": {
      "Scm": "git",
      "Name": "Devlead/Cake",
      "Branch": "krippz-bamboo-build",
      "Commit": {
        "RepositoryRevision": "a7c84e987b721b1832e5b7f8671e32d530639211"
      }
    }
  }
}
```

So looks good, so would be great if you could

1. Revert the change to `CakeEnvironment.cs`
2. Rebase against latest develop into 1 single commit
3. Remove WIP and change title to `Add support for Atlasssian Bamboo Build Server` and refer to Closing issue #544 in issue in commit/PR message (read more [here](https://help.github.com/articles/closing-issues-via-commit-messages/) ). 

and then I'll do a final review before merging.

krippz commented:
@devlead I did make some changes, I renamed BambooProjectInfo to BambooPlanInfo made more sense in the Bamboo concept. I also moved some stuff from the BambooEnvironmentInfo to BambooPlanInfo, made more sense to group these. 
Cheers

gep13 commented:
@krippz are you familiar with how to rebase your branch against the latest in the develop branch?  A rebase is the preferred option, rather than merging the changes, as it stops the history of the branches from getting polluted with merge commits, like the one that is currently in this PR.  If you aren't familiar with it, don't worry, we can talk you through the process.

krippz commented:
@gep13 rebased :)

gep13 commented:
>@krippz said...
@gep13 rebased :)

Perfect, thank you!

devlead commented:
@krippz merged, thanks for your contribution :+1:

patriksvensson commented:
Awesome :+1:
